### ISSUE TITLE
File format not recognized on .rlib

### ISSUE THREAD

**Issue by jeromegn** [alexcrichton/cc-rs#354]
Hey there, I originally opened the issue [in the rust repo](https://github.com/rust-lang/rust/issues/54971), but I think it might be related to this crate.

[This comment](https://github.com/japaric/rust-cross/issues/16#issuecomment-209686426) made me think it could be the way I'm compiling my C++ with this crate.

Here's the relevant lines in the build file: https://github.com/superfly/fly.rs/blob/master/libfly/build.rs#L35-L41

The error I'm getting is hard to debug:

```
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.15kq92zzbmxot4k9.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.16u6js6g0l3k1ic6.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.181cuta0v63atwcm.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.1c4sbqhvukbgthag.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.1im38lueib99jsk0.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.1kduva7sc7em934m.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.1mvmz58owquyropc.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.1vut2eft6nlujjxr.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.1y16o1qfye96o7m0.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.1zwd8n7bcl3vhvvh.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.23tqyymcb18u96mb.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.2imnh2vhxcqrizhm.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.2jqywn86b2gsqohu.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.2lyh15q6cjwzy18c.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.2qhkzqx5zqexj20y.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.3ayaeypdcro9d6yk.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.3py0d821mvt07s4n.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.3rngp6bm2u2q5z0y.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.3wta9ctgdrpkmlpr.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.44bsbddupzfao2om.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.48721dc4k5qxei0u.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.49a7n47po4ttqjl7.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.49lx1q7cxvpykyv0.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4b8ptp1vn215jmoe.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4ezmh1vbs95c5ack.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4oc10dk278mpk1vy.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4vp4wqj2v29i7mgy.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4xq48u46a1pwiqn7.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4ybye971cqflgun6.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4yh8x2b62dcih00t.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.4ypvbwho0bu5tnww.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.56dly8q07ws8ucdq.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.57qy3vyd9bhiuaon.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.5gf6du7k58s78kob.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.8xzrsc1ux72v29j.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.98g0d9x8aw3akpe.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.9elsx31vb4it187.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.c6lbtaiefvx3wya.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.oa3rad818d8sgn4.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.spyjbt69vcsrx9q.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.y08g5q2x813c4wx.rcgu.o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.z9ox7biyn1otfln.rcgu.o" "-o" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba" "/usr/src/myapp/target/debug/deps/create_snapshot-60efbce147cacbba.crate.allocator.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "/usr/src/myapp/target/debug/deps" "-L" "/usr/src/myapp/target/debug/build/libfly-2ef616123afaf643/out" "-L" "/usr/src/myapp/libfly/third_party/v8/out.gn/x64.release/obj" "-L" "/usr/src/myapp/target/debug/build/libsqlite3-sys-7ac6aaa28420bed7/out" "-L" "/usr/src/myapp/target/debug/build/backtrace-sys-dc873d7131d9513f/out" "-L" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/usr/src/myapp/target/debug/deps/liblibfly-973820a835f9f432.rlib" "/usr/src/myapp/target/debug/deps/liblibc-40d568469a8c2e87.rlib" "-Wl,--start-group" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-9318d1aa9575dbf9.rlib" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-e3cd3f44688b2d97.rlib" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-2f4890fbea3bd5e0.rlib" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-a0ddde720c2c46c5.rlib" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_system-9c41ffe739844496.rlib" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-af766b046896c123.rlib" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-00b776688b98de66.rlib" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-d05a4396ceff8bc8.rlib" "-Wl,--end-group" "/usr/local/rustup/toolchains/1.29.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-fe16a4dcdcd95bab.rlib" "-Wl,-Bdynamic" "-lstdc++" "-lutil" "-lutil" "-ldl" "-lrt" "-lpthread" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
  = note: /usr/src/myapp/target/debug/deps/liblibfly-973820a835f9f432.rlib: error adding symbols: File format not recognized
         collect2: error: ld returned 1 exit status
```

Any idea? Could it be related to a different lib (V8) which I'm statically linking? As far as I can tell, the linking works fine and I usually get errors earlier in the build when that's wrong.

A similar project ([deno](/denoland/deno)) uses `gn` to build both the rust crates and V8 "together". Hopefully that's not the solution as this build process is much more involved and I'm not very familiar with it.

Would appreciate any help!

**Comment by alexcrichton** [comment]
Thanks for the report! I can try to help dig in here, but I'm not currently sure how to build this. Do you have some pointers for how to get the build up and running to poke around this locally?

**Comment by jeromegn** [comment]
That could be part of the problem. I'm not sure if using a pre-built V8 is at fault here. It's somewhat a requirement though: it can take up to 30 mins to compile and uses about 15GB of disk space to do so.

The [fly.rs repository](https://github.com/superfly/fly.rs)'s README has some information for building on macos, but for linux, it's best to inspire from the Dockerfile.

AFAICT, the same issue arises when building either in a docker or non-docker environment.

The build process is a bit rough, if you can use docker, that's ideal. This [Dockerfile](https://github.com/superfly/fly.rs/blob/master/Dockerfile) should do the trick.

If not, you need:
- V8 built in `libfly/third_party/v8/out.gn/x64.release`. You can extract a pre-compiled version from the docker image [`flyio/v8:7.1`](https://hub.docker.com/r/flyio/v8/) based on Debian stretch, built for x64 
- `flatc` in your $PATH (built from google/flatbuffers) but only certain versions work (like the one checked out in `third_party/`.)
- `yarn` (package manager for node.js). You may not need this one if building in debug mode, but release mode will embed some files.

Hopefully this isn't too scary. It's not the easiest project to build, that's for sure. We only have 2 people working on it right now, we'll make the process smoother soon!

Small update on what I tried:
- Switching `cc` to use `clang` instead of `gcc` (no change, I wiped the target directories before trying)
- Using `ld.gold` linker instead of `ld` (that was a disaster, lots of missing symbols)

Appreciate any help you can offer on this.

**Comment by alexcrichton** [comment]
Oh that's perfect! 

So digging into this, it looks like `libfly` in Rust statically links to `v8_monolith`, and it looks like that library is compiled with LTO. All of its object files aren't actually object files but are instead LLVM IR. Is there a way to disable LTO for v8 that you're building? 

The error here is happening because the linker (presumably?) doesn't have the right flags passed to it to enable link-time-code-generation (LTO), and `ld` doesn't understand LLVM IR, so it's failing at link time.

**Comment by jeromegn** [comment]
Ohhh. That makes sense.

I've been pre-building V8, but omitted a script from the original source. Would something like that: https://github.com/augustoroman/v8/blob/master/docker-v8-lib/fatten_archives.sh help to make them not thin? I don't know enough about how all this works.

I'm sure there's an option somewhere to disable LTO though. I'll look into that.

**Comment by jeromegn** [comment]
This looks like just the thing: https://groups.google.com/forum/#!searchin/v8-users/lto%7Csort:date/v8-users/MAJwyq2WEQQ/SzRmylHzBQAJ

**Comment by jeromegn** [comment]
New errors, which is good.

https://dev.azure.com/flydotio/fly.rs/_build/results?buildId=180&view=logs (clicking "CmdLine" should expand the logs)

I attempted to fix these by explicitly linking stdc++ in *both* `libfly` and the main (`fly`) crate, but I still get these errors.

I didn't have to link to stdc++ under macos though.

Example error:

```
2018-10-12T20:32:42.8413989Z           /build/chromium/v8/out.gn/lib/../../buildtools/third_party/libc++/trunk/include/memory:3569: undefined reference to `std::__1::__shared_weak_count::__release_weak()'
```

**Comment by alexcrichton** [comment]
Hm I think that may also mean that the wrong libstdc++ is linked in? I think that Clang's is slightly different in terms of ABI and may be called libc++ instead of libstdc++? Does that make a difference perhaps?

**Comment by jeromegn** [comment]
I've been able to compile by link to libc++. Thanks!

My binary went from 30MB to about 500MB though. I suppose LTO helped a lot there. I'll have to find a way to reduce the file size here.

I'm not sure why it would work fine on macOS with LTO enabled and produce such lean binaries.

**Comment by alexcrichton** [comment]
Ok great! I'm gonna go ahead and close this as it looks like it's largely been solved. It may be the case that v8 doesn't play nicely with `--gc-sections` and relies on LTO to remove unneeded functionality, although I'm not sure :(

**Comment by jeromegn** [comment]
Interesting. Closing this is :+1:

As an update: The only arg I had to use was `is_cfi = false` which is this: https://clang.llvm.org/docs/ControlFlowIntegrity.html which:

> To enable Clang’s available CFI schemes, use the flag -fsanitize=cfi. You can also enable a subset of available schemes. As currently implemented, all schemes rely on link-time optimization (LTO); so it is required to specify -flto, and the linker used must support LTO, for example via the gold plugin.

I was able to go back under 30MB by stripping the binary (from 500MB down to 30MB is still surprising...) I'm not entirely sure why I don't have to do that on macOS.

It's unclear to me what the side-effects of stripping the binary might have. I assume it'll be harder to debug in case of a crash.

I've been trying out the tricks from: https://lifthrasiir.github.io/rustlog/why-is-a-rust-executable-large.html
